Sources/GRPC/ConnectionPool/PoolManager.swift

```diff
@@ -0,0 +1,270 @@
+      // call `initialize` directly.
+      self.state.activate(pools: pools, assumingPerPoolCapacity: assumedCapacity)
+    }
+
+    for pool in pools {
+      pool.initialize(connections: configuration.connections)
+    }
+  }
+
+  /// Make one pool per `EventLoop` in the pool's `EventLoopGroup`.
+  /// - Parameters:
+  ///   - configuration: The configuration to make each pool with.
+  ///   - logger: A logger.
+  /// - Returns: An array of `ConnectionPool`s.
+  private func makePools(
+    perPoolConfiguration configuration: PerPoolConfiguration,
+    logger: GRPCLogger
+  ) -> [ConnectionPool] {
+    let eventLoops = self.group.makeIterator()
+    return eventLoops.map { eventLoop in
+      // We're creating a retain cycle here as each pool will reference the manager and the per-pool
+      // state will hold the pool which will in turn be held by the pool manager. That's okay: when
+      // the pool is shutdown the per-pool state and in turn each connection pool will be dropped.
+      // and we'll break the cycle.
+      return ConnectionPool(
+        eventLoop: eventLoop, maxWaiters: configuration.maximumWaiters,
+        reservationLoadThreshold: configuration.loadThreshold,
+        assumedMaxConcurrentStreams: configuration.assumedMaxConcurrentStreams,
+        channelProvider: configuration.channelProvider,
+        streamLender: self,
+        logger: logger
+      )
+    }
+  }
+
+  // MARK: Stream Creation
+
+  /// Make a stream and initialize it.
+  ///
+  /// - Parameters:
+  ///   - preferredEventLoop: The `EventLoop` that the stream should be created on, if possible. If
+  ///       a pool exists running this `EventLoop` then it will be chosen over all other pools,
+  ///       irregardless of the load on the pool. If no pool exists on the preferred `EventLoop` or
+  ///       no preference is given then the pool with the most streams available will be selected.
+  ///       The `EventLoop` of the selected pool will be the same as the `EventLoop` of
+  ///       the `EventLoopFuture<Channel>` returned from this call.
+  ///   - deadline: The point in time by which the stream must have been selected. If this deadline
+  ///       is passed then the returned `EventLoopFuture` will be failed.
+  ///   - logger: A logger.
+  ///   - initializer: A closure to initialize the `Channel` with.
+  /// - Returns: An `EventLoopFuture<Channel>` on the same `EventLoop` as the connection pool
+  ///     providing the `Channel`. The future will be failed if the pool manager has been shutdown,
+  ///     the deadline has passed before a stream was created or if the selected connection pool
+  ///     is unable to create a stream (if there is too much demand on that pool, for example).
+  internal func makeStream(
+    preferredEventLoop: EventLoop?,
+    deadline: NIODeadline,
+    logger: GRPCLogger,
+    streamInitializer initializer: @escaping (Channel) -> EventLoopFuture<Void>
```

Comment:
This API is likely to be somewhat hard to hold: the future can run on an essentially arbitrary event loop, making chaining from it difficult. We should consider either using a `promise` or requiring an event loop for the returned future to execute on.

Reply:
I agree it's slightly awkward but I'm not sure how to work around it best. The problem is that if the caller doesn't have an event loop preference then the event loop of the selected pool (and returned future) dictate which event loop the RPC should run on.

I suppose one option would be to more explicitly indicate which event loop it's running on:

```swift
enum SelectedStreamEventLoop {
  case preferredEventLoop(EventLoopFuture<Channel>)
  case other(EventLoopFuture<Channel>)
}
```

Although I'm not sure how much this actually adds since that's information already encoded by the future.

Reply:
If this is load-bearing then I think this needs to call be more explicit, yeah. That information is very _implicit_ in the future. I think we shouldn't just return the future but the loop as well, to make it very clear that we're trying to communicate "this is the loop on which the RPC will run".